File: c++/sortlist.cpp
// Sort a linked list using insertion sort.

#include <iostream>
using namespace std;

// * Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
 
/*
ListNode* mergeSortList(ListNode *head) {
    ListNode *p = head;
    int size = 0;
    while (p) {
        size++;
        p = p->next;
    }
    ListNode* result = mergesort(head, size);
    return result;
}

ListNode* mergesort(ListNode* head, int size) {
    if (size > 1) {
        ListNode* l1 = head;
        int l1size = size / 2;
        int cnt = 0;
        ListNode* l2 = head;
        while (cnt < l1size)  {
            cnt++;
            l2 = l2->next;
        }
        int l2size = size -l1size;
        mergesort(l1, l1size);
        mergesort(l2, l2size);
        ListNode* result = merge_list(l1, l1size, l2, l2size);
        return result;
    }
    return head;
}

ListNode* merge_list(ListNode* l1, int l1size, ListNode* l2, int l2size) {
    int i, j, tmp;
    i = j = 0;
    while (i < l1size && j < l2size) {
        if (*l1 > *l2) {
            tmp = *l1;
            *l1 = *l2;
            *l2 = tmp;
            i++;
            l1 = l1->next;
        } else {


        }


    }

}
*/
ListNode* init(int a[], int n) {
    ListNode* head = new ListNode(a[0]);
    ListNode *curr = head;

    for (int i = 1; i < n; ++i) {
        ListNode* tmp = new ListNode(a[i]);
        curr->next = tmp;
        curr = curr->next;
    }
    return head;
}

void printList(ListNode *head) {
    if (head == NULL) return;
    ListNode * tmp = head;
    while (tmp != NULL) {
        cout << tmp->val << ", ";
        tmp = tmp->next;
    }
    cout << endl;
}

ListNode* sortList(ListNode* head) {
    if (!head || !head->next) return head;
    ListNode *mid = head, *tmp = head->next;
    while (tmp && tmp->next) {  // slow fast ptrs
        mid = mid->next;
        tmp = tmp->next->next;
    }
    tmp = mid->next;  // pointing to head of second half
    mid->next = NULL; // cut from the middle
    ListNode *l1 = sortList(head);
    ListNode *l2 = sortList(tmp);
    
    // insert l2 to l1
    ListNode **curr = &l1;
    ListNode *entry;
    while (l2 && *curr) {
        entry = *curr;
        if (l2->val < entry->val) { // l1->val < l2->val
            *curr = l2;
            l2 = entry;
        }
        // l1->val > l2->val
        curr = &(*curr)->next;
    }
    if (l2) *curr = l2;
    return l1;
}

int main() {
    int a[] = {2, 1, 3, 7};
    int n = 4;
    ListNode *head = init(a, n);
    printList(head);

    ListNode *p = sortList(head);
    printList(p);
    return 0;
}

File: java/sortList.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class sortList {
    public static class Solution {
        public class ListNode {
            int val;
            ListNode next;
            ListNode(int x) {
                val = x;
                next = null;
            }
        }

        public ListNode ListNode(int x) {
            ListNode res = new Solution.ListNode(-1);
            res.val = x;
            res.next = null;
            return res;
        }

        public ListNode divideList(ListNode head) {
            ListNode fast = head;
            ListNode slow = head;
            ListNode prev = head;
            while (fast != null && fast.next != null) {
                prev = slow;
                slow = slow.next;
                fast = fast.next.next;
            }
            prev.next = null;
            return slow;
        }
        
        public ListNode mergeList(ListNode one, ListNode two) {
            if (one == null) return two;
            if (two == null) return one;
            ListNode dummy = new ListNode(Integer.MIN_VALUE);
            ListNode curr = dummy;
            while (one != null && two != null) {
                if (one.val < two.val) {
                    curr.next = one;
                    one = one.next;
                } else {
                    curr.next = two;
                    two = two.next;
                }
                curr.next.next = null;
                curr = curr.next;
            }
            if (one != null) curr.next = one;
            else curr.next = two;
            return dummy.next;
        }
        
        public ListNode sortList(ListNode head) {
            if (head == null || head.next == null) {
                return head;
            }
            ListNode second = divideList(head);
            ListNode firstSorted = sortList(head);
            ListNode secondSorted = sortList(second);
            return mergeList(firstSorted, secondSorted);
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a= {6, 5, 3, 1, 2};
        Solution.ListNode one = result.ListNode(a[0]);
        Solution.ListNode curr = one;
        for (int i = 1; i < a.length; i++) {
            Solution.ListNode tmp = result.ListNode(a[i]);
            curr.next = tmp;
            curr = curr.next;
        }

        curr = result.sortList(one);
        while (curr != null) {
            System.out.println(curr.val);
            curr = curr.next;
        }
    }
}
